rllib/algorithms/sac/sac.py

```diff
@@ -371,3 +430,133 @@ def get_default_policy_class(
+                        lambda w: w.sample(),
+                        local_worker=False,
+                    )
+                # Perform SAC postprocessing (Prio weights) on a (flattened)
+                # list of Episodes.
+                # TODO (simon): Change this to a preprocessing in the learner.
+                # postprocessed_episodes = self.postprocess_episodes(episodes)
+
+            episodes = tree.flatten(episodes)
+            # TODO (sven): single- vs multi-agent.
+            self._counters[NUM_AGENT_STEPS_SAMPLED] += sum(len(e) for e in episodes)
+            self._counters[NUM_ENV_STEPS_SAMPLED] += sum(len(e) for e in episodes)
+
+            # Add the sampled experiences to the replay buffer.
+            # TODO (simon): Apply the same function like in PPO to aggregate episodes
+            # over workers.
+            self.local_replay_buffer.add(episodes)
+
+        # Update the target network each `target_network_update_freq` steps.
+        current_ts = self._counters[
+            NUM_AGENT_STEPS_SAMPLED
+            if self.config.count_steps_by == "agent_steps"
+            else NUM_ENV_STEPS_SAMPLED
+        ]
+
+        # If enough experiences have been sampled start training.
+        if current_ts > self.config.num_steps_sampled_before_learning_starts:
+            # Run multiple training iterations.
+            for _ in range(sample_and_train_weight):
+                # Sample training batch from replay_buffer.
+                train_dict = self.local_replay_buffer.sample(
+                    num_items=self.config.train_batch_size,
+                    # TODO (simon): Implement for n-step adjustment.
+                    # batch_length_T=self.config.n_step + 1,
+                    batch_length_T=2,
+                )
+
+                # TODO (sven): This looks a bit ugly, but we have not the same
+                # naming in `SingleAgentEpisode` and `SampleBatch`.
+                # Note, the `SingleAgentEpisode` stores obs_t. action_(t-1),
+                # reward_t, infos_t, while the replay buffer's `sample()`
+                # returns obs_t, action_t, reward_(t-1). So, we take the
+                # reward from the subsequent experience in the episode to
+                # calculate the Q-values.
+                train_batch = SampleBatch(
+                    {
+                        SampleBatch.OBS: train_dict[SampleBatch.OBS][:, 0],
+                        SampleBatch.NEXT_OBS: train_dict[SampleBatch.OBS][:, 1],
+                        SampleBatch.ACTIONS: train_dict[SampleBatch.ACTIONS][:, 0],
+                        SampleBatch.REWARDS: train_dict[SampleBatch.REWARDS][:, 1],
+                        SampleBatch.TERMINATEDS: train_dict["is_terminated"][:, 0],
+                        SampleBatch.TRUNCATEDS: train_dict["is_truncated"][:, 0],
+                        "is_first": train_dict["is_first"][:, 0],
```

Comment:
I don't think we need these flags here (I believe only DreamerV3 does for now).

Reply:
What is with the Q target when the first timestep in the tuple is already the last one. Then the second one comes from the beginning of another episode, correct? 

I used for the masked target Q `(1.0 - batch[SampleBatch.TERMINATEDS]) * (1.0 - batch["is_last"]) * q_target_next`

Reply:
We just have to be careful: "is_last" does NOT mean the episode is done. If "is_last" is True, it could mean:
* we simply are at the end of a batch row (of len=T), but still in the middle of an episode.
* we are in the middle of a batch row, but the episode was truncated (not terminated!). Hence we should NOT set the q-value target to 0.0 at these locations (but would need to compute the actual Q-values for the observations found there).

But again, I don't think we should use this buffer class, but subclass it and override the `sample()` method. We can definitely reuse the EpisodeReplayBuffer's handling of adding episode chunks, but its sampling method would not be suitable for SAC, imo.